src/ram/transform/ChoiceConversion.cpp

```diff
@@ -105,6 +120,37 @@ Own<Operation> ChoiceConversionTransformer::rewriteIndexScan(const IndexScan* in
         return mk<IndexChoice>(rel, identifier, souffle::clone(filter->getCondition()), std::move(newValues),
                 souffle::clone(filter->getOperation()), indexScan->getProfileText());
     }
+
+    // Check that Relation is not referenced further down in the loop nest
+    bool referencedBelow = false;
+    visitDepthFirst(*indexScan, [&](const TupleElement& element) {
+        if (element.getTupleId() == indexScan->getTupleId()) {
+            referencedBelow = true;
+        }
+    });
+
+    // Convert the Scan into a Choice where True
+    if (!referencedBelow) {
+        RamPattern newValues;
+        for (auto& cur : indexScan->getRangePattern().first) {
+            Expression* val = nullptr;
```

Comment:
Hi, 

Is it actually necessary to insert the `nullptr`s into newValues here?  I've spent a fair bit of time in the ast refactor to ensure that everything in the tree is valid, and it would be, IMHO, best to try to accomplish the same thing with the RAM nodes.  It would be nicer to avoid having to perform nullptr checks in the code.

I would also suggest to avoid using naked `clone`; again, I've already gone through this conversion in `ast` but since @azreika was working on RAM, I didn't touch that part of the codebase.  Together, my suggestion would be to try write something akin to

```suggestion
            if (cur != nullptr) {
                newValues.first.emplace_back(souffle::clone(*cur));
            }
```
which strikes me as a bit cleaner, and it's also exception-safe. :smile: